packages/graphql/lib/src/link/link.dart

```diff
@@ -12,26 +12,42 @@ typedef RequestHandler = Stream<FetchResult> Function(
   NextLink forward,
 ]);
 
-Link _concat(
-  Link first,
-  Link second,
-) {
-  return Link(request: (
-    Operation operation, [
-    NextLink forward,
-  ]) {
-    return first.request(operation, (Operation op) {
-      return second.request(op, forward);
-    });
-  });
+abstract class Link {
+  static Link from(List<Link> links) => links.reduce(
+        (first, next) => first.concat(next),
+      );
+
+  static Link fromHandler(RequestHandler handler) =>
+      _RequestHandlerLink(handler);
+
+  Stream<FetchResult> request(Operation operation, [NextLink forward]);
+
+  Link concat(Link next) => _ConcatLink(this, next);
+}
+
+class _RequestHandlerLink extends Link {
+  _RequestHandlerLink(this._handler);
+
+  RequestHandler _handler;
+
+  @override
+  Stream<FetchResult> request(Operation operation, [NextLink forward]) {
+    return _handler(operation, forward);
+  }
 }
 
-class Link {
-  Link({this.request});
+class _ConcatLink extends Link {
```

Comment:
This seems like overkill - let's stick to `_concat`. What `concat` really needs is [more safety / warnings](https://github.com/apollographql/apollo-link/blob/089f71be5ea6a2ccd465751a500b16b97dbc23aa/packages/apollo-link/src/link.ts#L62)